src/core/qgstessellator.cpp

```diff
@@ -436,21 +459,66 @@ void QgsTessellator::addPolygon( const QgsPolygon &polygon, float extrusionHeigh
 
   if ( pCount == 4 && polygon.numInteriorRings() == 0 )
   {
+    std::vector<p2t::Point *> polyline;
+    if ( mAddTextureCoords )
+    {
+      std::unique_ptr<QMatrix4x4> toNewBase, toOldBase;
+      if ( !mNoZ && pNormal != QVector3D( 0, 0, 1 ) )
+      {
+        // this is not a horizontal plane - need to reproject the polygon to a new base so that
+        // we can do the triangulation in a plane
+
+        QVector3D pXVector, pYVector;
+        _normalVectorToXYVectors( pNormal, pXVector, pYVector );
+
+        // so now we have three orthogonal unit vectors defining new base
+        // let's build transform matrix. We actually need just a 3x3 matrix,
+        // but Qt does not have good support for it, so using 4x4 matrix instead.
+        toNewBase.reset( new QMatrix4x4(
+                           pXVector.x(), pXVector.y(), pXVector.z(), 0,
+                           pYVector.x(), pYVector.y(), pYVector.z(), 0,
+                           pNormal.x(), pNormal.y(), pNormal.z(), 0,
+                           0, 0, 0, 0 ) );
+
+        // our 3x3 matrix is orthogonal, so for inverse we only need to transpose it
+        toOldBase.reset( new QMatrix4x4( toNewBase->transposed() ) );
+      }
```

Comment:
It would be good to move the rotation code to a new function to avoid code duplication...